_posts/2023-02-19-a-lean-dataops-platform.md

```diff
@@ -0,0 +1,158 @@
+
+is_resource_grant(elem) = true {
+  arr[_] = elem
+} else = false { true }
+
+
+planned_schema_grants_resources = [ res |
+  res := input.resource_changes[_]
+  is_resource_grant(res.type)
+]
+
+deny[msg] {
+  planned_schema_grants_resources[i].change.after.enable_multiple_grants == false
+  msg := sprintf("\n\tFor resource: %v \n\tParameter enable_multiple_grants set to: %v, this should be true",[planned_schema_grants_resources[i].address, planned_schema_grants_resources[i].change.after.enable_multiple_grants])
+}
+
+```
+
+This runs as one of our safety checks within our pull requests in github to ensure no one's access is revoked unexpectedly.
+
+In addition to this issue, there is some difficulty in troubleshooting errors as sometimes the error message is a red-hearing or is not descriptive enough to understand what went wrong. My personal method of troubleshooting these errors is to reset the terraform state of specific objects and check the Snowflake query logs to understand what the provider was trying to do.
+
+## Fivetran 
+Similar to the Snowflake Terraform provider, the Fivetran Terraform provider is also in its infancy. The main drawback of this is that some of the functionality available in the Fivetran console has not been brought over to the Terraform provider just yet, however, from discussions with the team at Fivetran, they are working hard to make this a reality so stay tuned!
+
+## Dagster
+The integrations in Dagster between each service are ok, fitting our use case but unfortunately stripping us of some control. An example of this is if a Fivetran sync fails silently its status is set to "rescheduled". In these cases, typically following a historical sync, the "rescheduled" sync never occurs. This causes Dagster to poll indefinitely until a timeout occurs. Unfortunately, I have not found a way to resolve this error in Dagster and thus it would be nice to have the ability to catch this polling message however Dagster doesn't allow us to do this. Though I want to emphasize this is an issue with Fivetran not Dagster. 
+# Conclusion
+A solid foundation of quality, aggregated, and easy-to-access data is necessary to even start making decisions that are data-driven. To achieve this, companies need to leverage the best tools available and apply good engineering practices such as data-ops coding practices. The internal data platform we built by leveraging multiple greenfield technologies, such as Snowflake, Fivetran, and DBT. They also utilized the Dagster orchestration tool and created a CI/CD pipeline using Github actions. While these tools have some drawbacks, they allowed us to create a lean data platform with robust security, governance checks, and data testing. The platform is still in its early iteration, and it will likely change in the future. Nonetheless, the adoption of data-ops practices and the use of modern technologies enabled us to achieve our key objectives and move closer to becoming data-driven.
```

Comment:
the adoption of data-ops practices and the use of modern technologies enabled us to achieve our key objectives

this is really a case of dogfooding (ie. practicing internally what we preach to customers)